### PR TITLE
Adding test of Gwen's activity

### PR DESCRIPTION
**Pull request by gkirschke**
Example files from class; will update later

### CODE REVIEW

File: activities/Gwen.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+      "scanning_params": {
+        "smooth": 10,
+        "zexag": 2,
+        "numscans": 1,
+        "interpolate": true
+      },
+      "analyses": "Gwen.py",
+      "title": "Selecting locations for pollinator gardens that minimize overall travel time between floral resources",
+      "author": "Gwen Kirschke",
+      "instructions": "Add garden locations and program will calculate minimum distance between gardens"
```
Comment (chaedri) [inline activities/Gwen.json]: So cool!!!

File: activities/Gwen.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    max_detected,
+    debug,
+    env,
+):
+    diff_thr = "diff_thr_" + str(uuid.uuid4()).replace("-", "")
+    diff_thr_clump = "diff_thr_clump_" + str(uuid.uuid4()).replace("-", "")
+    coeff = gcore.parse_command(
+        "r.regression.line", mapx=after, mapy=before, flags="g", env=env
+    )
+    grast.mapcalc(
```
Comment (chaedri) [inline activities/Gwen.py]:
```suggestion
    gs.mapcalc(
```

File: activities/Gwen.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    height_threshold,
+    cells_threshold,
+    add,
+    max_detected,
+    debug,
+    env,
+):
+    diff_thr = "diff_thr_" + str(uuid.uuid4()).replace("-", "")
+    diff_thr_clump = "diff_thr_clump_" + str(uuid.uuid4()).replace("-", "")
+    coeff = gcore.parse_command(
```
Comment (chaedri) [inline activities/Gwen.py]:
```suggestion
    coeff = gs.parse_command(
```

File: activities/Gwen.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    )
+    grast.mapcalc(
+        "diff = {a} + {b} * {after} - {before}".format(
+            a=coeff["a"], b=coeff["b"], before=before, after=after
+        ),
+        env=env,
+    )
+    try:
+        if add:
+            grast.mapcalc(
```
Comment (chaedri) [inline activities/Gwen.py]:
```suggestion
            gs.mapcalc(
```

File: activities/Gwen.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                    diff_thr=diff_thr,
+                    after=after,
+                    before=before,
+                    thr1=height_threshold[0],
+                    thr2=height_threshold[1],
+                ),
+                env=env,
+            )
+        else:
+            grast.mapcalc(
```
Comment (chaedri) [inline activities/Gwen.py]:
```suggestion
            gs.mapcalc(
```

File: activities/Gwen.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+                    a=coeff["a"],
+                    b=coeff["b"],
+                    after=after,
+                    before=before,
+                    thr=height_threshold,
+                ),
+                env=env,
+            )
+
+        gcore.run_command("r.clump", input=diff_thr, output=diff_thr_clump, env=env)
```
Comment (chaedri) [inline activities/Gwen.py]:
```suggestion
        gs.run_command("r.clump", input=diff_thr, output=diff_thr_clump, env=env)
```

File: activities/Gwen.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        stats = (
+            gcore.read_command(
```
Comment (chaedri) [inline activities/Gwen.py]:
```suggestion
            gs.read_command(
```

File: activities/Gwen.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+                if (
+                    float(stat.split()[1]) < cells_threshold[1]
+                    and float(stat.split()[1]) > cells_threshold[0]
+                ):  # larger than specified number of cells
+                    found += 1
+                    cat, value = stat.split()
+                    cats.append(cat)
+            if cats:
+                rules = ["{c}:{c}:1".format(c=c) for c in cats]
+                gcore.write_command(
```
Comment (chaedri) [inline activities/Gwen.py]:
```suggestion
                gs.write_command(
```

File: activities/Gwen.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+                    "r.recode",
+                    input=diff_thr_clump,
+                    output=change,
+                    rules="-",
+                    stdin="\n".join(rules),
+                    env=env,
+                )
+                gcore.run_command(
```
Comment (chaedri) [inline activities/Gwen.py]:
```suggestion
                gs.run_command(
```

File: activities/Gwen.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+                    "r.volume",
+                    flags="f",
+                    input=change,
+                    clump=diff_thr_clump,
+                    centroids=change,
+                    env=env,
+                )
+            else:
+                gcore.warning("No change found!")
```
Comment (chaedri) [inline activities/Gwen.py]:
```suggestion
                gs.warning("No change found!")
```

File: activities/Gwen.py [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+                gcore.run_command("v.edit", map=change, tool="create", env=env)
```
Comment (chaedri) [inline activities/Gwen.py]:
```suggestion
                gs.run_command("v.edit", map=change, tool="create", env=env)
```

File: activities/Gwen.py [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+        else:
+            gcore.warning("No change found!")
```
Comment (chaedri) [inline activities/Gwen.py]:
```suggestion
            gs.warning("No change found!")
```

File: activities/Gwen.py [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+            gcore.run_command("v.edit", map=change, tool="create", env=env)
```
Comment (chaedri) [inline activities/Gwen.py]:
```suggestion
            gs.run_command("v.edit", map=change, tool="create", env=env)
```

File: activities/Gwen.py [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+
+        gcore.run_command(
```
Comment (chaedri) [inline activities/Gwen.py]:
```suggestion
        gs.run_command(
```

File: activities/Gwen.py [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+            "g.remove",
+            flags="f",
+            type=["raster"],
+            name=[diff_thr, diff_thr_clump],
+            env=env,
+        )
+    except:
+        gcore.run_command(
```
Comment (chaedri) [inline activities/Gwen.py]:
```suggestion
        gs.run_command(
```

File: activities/Gwen.py [hunk 16]
```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env python3
+
+import os
+
+import grass.script as gs
+
+import numpy as np
+
```
Comment (chaedri) [inline activities/Gwen.py:7]: Found this in tangible_utils and you use it later.

```suggestion
def get_environment(**kwargs):
    """!Returns environment for running modules.
    All modules for which region is important should
    pass this environment into run_command or similar.

    @param tmp_regions a list of temporary regions
    @param kwargs arguments for g.region

    @return environment as a dictionary
    """
    env = os.environ.copy()
    env["GRASS_OVERWRITE"] = "1"
    env["GRASS_VERBOSE"] = "0"
    env["GRASS_MESSAGE_FORMAT"] = "standard"
    region3d = False
    if "raster_3d" in kwargs:
        region3d = True
    env["GRASS_REGION"] = gs.region_env(region3d=region3d, **kwargs)
    return env

```

File: activities/Gwen.py [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+        gcore.run_command(
+            "g.remove",
+            flags="f",
+            type=["raster"],
+            name=[diff_thr, diff_thr_clump],
+            env=env,
+        )
+
+def run_distance(real_elev, scanned_elev, env, **kwargs):
+    env2 = get_environment(raster=scanned_elev, n='n-100', s='s+100', e='e-100', w='w+100') # not sure what shifting this does?
```
Comment (chaedri) [inline activities/Gwen.py]: This shifting seems to be making the computational region smaller (removing a buffer of 100 on all sides). @petrasovaa might have an idea why this is necessary.

File: activities/Gwen.json [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+          "d.vect", "map=data", "color=red", "fill_color=orange",
+          "icon=basic/star", "size=15"
+        ]
+      ],
+      "base": "elev_lid792_1m",
+      "scanning_params": {
+        "smooth": 10,
+        "zexag": 2,
+        "numscans": 1,
+        "interpolate": true
```
Comment (chaedri) [inline activities/Gwen.json:23]:
```suggestion
        "interpolate": true,
        "calibrate":true
```

File: activities/Gwen.py [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+def run_distance(real_elev, scanned_elev, env, **kwargs):
+    env2 = get_environment(raster=scanned_elev, n='n-100', s='s+100', e='e-100', w='w+100') # not sure what shifting this does?   
+    before = 'scan_saved' #shouldn't this be real_elev?
+    
+    # get new garden location points from pins
+    change_detection(before=before, after=scanned_elev,
+                              change='change', height_threshold=[15, 100], cells_threshold=[5, 100], add=True, max_detected=13, debug=True, env=env2)
+    # dict of new points
+    points = {}
+    data = gs.read_command('v.out.ascii', input='trail_points', type='point', format='point', env=env).strip() # where does trail_points come from?
```
Comment (chaedri) [inline activities/Gwen.py]: `trail_points` is the A and B points from that trail exercise that already exist. A little future down in the trails code, you can see the points get extracted from the change vector: https://github.com/tangible-landscape/tangible-landscape-applications/blob/4a63c3ae673139fc29ba3a92ce9acbe51bedbcd5/planting/workflows/trails.py#L41

In other words, delete this block of code and replace it with something like:

```
     points = {}
    gs.run_command('v.edit', tool='create', map=trail, env=env)
    # detected points
    points_raw = gscript.read_command('v.out.ascii', input='change',
                                      type='point', format='point').strip().split()
    for point in points_raw:
        point = point.split('|')
        point = (float(point[0]), float(point[1]))
        points[i] = point
        i += 1
```

File: activities/Gwen.py (hunk 17 above)
Comment (petrasovaa) [reply activities/Gwen.py]: That's for practical reasons when scanning, the edges of the model can get messy. But you can remove that.

**Review by petrasovaa — COMMENTED** [review]
Why don't you use the advanced_example.py, that was meant as a starting point for tasks that involve marker detection.

File: activities/Gwen.json (hunk 1 above)
Comment (petrasovaa) [reply activities/Gwen.json]: How do you define minimum distance between gardens?

**Comment by petrasovaa** [comment]
To include a dashboard, check out https://github.com/tangible-landscape/grass-tangible-landscape/blob/master/activities_config_documentation.md#additional-widgets

**Comment by petrasovaa** [comment]
Looks like your json is invalid and the python file has a syntax error as well.

**Comment by gkirschke** [comment]
I'm not sure what the issue is with the json file; I thought that what I'd
added for the dashboard was the problem, but I removed that and it still
failed. I ran the python file through black locally; what else can I do to
identify or fix the remaining black failures?

--


Gwen Kirschke

Ph.D. Student, Geospatial Analytics | Center for Geospatial Analytics
| Youngsteadt
Urban Ecology Lab


Jordan Hall 5117 | Campus Box 7106

North Carolina State University

2800 Faucette Dr.

Raleigh, NC 27695 USA

***@***.*** | geospatial.ncsu.edu



"All electronic mail messages in connection with State business which are
sent to or received by this account are subject to the NC Public Records
Law and may be disclosed to third parties.”

**Comment by petrasovaa** [comment]
Click at Code Quality / Black details:
```
+++ /home/runner/work/gis714-2024-tangible-landscape/gis714-2024-tangible-landscape/activities/Gwen.py	2024-04-11 17:37:27.843997+00:00
@@ -82,13 +82,11 @@
     env["GRASS_OVERWRITE"] = "1"
     elevation = "elev_lid792_1m"
     elev_resampled = "elev_resampled"
     # We use resampling to get a similar resolution as with Tangible Landscape.
     gs.run_command("g.region", raster=elevation, res=4, flags="a", env=env)
-    gs.run_command(
-        "r.resamp.stats", input=elevation, output=elev_resampled, env=env
-    )
+    gs.run_command("r.resamp.stats", input=elevation, output=elev_resampled, env=env)
     # The end of the block which needs no editing.
 
     # Code specific to testing of the analytical function.
     # Create points which is the additional input needed for the process.
     points = "points"
@@ -100,12 +98,10 @@
         separator="comma",
         stdin="638[43](https://github.com/ncsu-geoforall-lab/gis714-2024-tangible-landscape/actions/runs/8650940174/job/23720512664?pr=3#step:3:45)2,220382\n638621,220607",
         env=env,
     )
     # Call the analysis.
-    run_function_with_points(
-        scanned_elev=elev_resampled, env=env, points=points
-    )
+    run_function_with_points(scanned_elev=elev_resampled, env=env, points=points)
```
and fix it accordingly.

JSON file is invalid, use an online json validator to see the problem.

File: activities/Gwen.py [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+        input="cfactorgrow_1m",
+        output="cost_surface",
+        start_points="focal_garden",
+        stop_points=points,
+    )
+
+    # calculate connectivity metric
+    data = gs.read_command("r.what", map="cost_surface", points=points)
+    data_lines = data.splitlines()
+    data = data_lines.split("|")
```
Comment (petrasovaa) [inline activities/Gwen.py]: From continuous integration failure:
```
Traceback (most recent call last):
  File "/home/runner/work/gis714-2024-tangible-landscape/gis714-2024-tangible-landscape/activities/Gwen.py", line 111, in <module>
    main()
  File "/home/runner/work/gis714-2024-tangible-landscape/gis714-2024-tangible-landscape/activities/Gwen.py", line 105, in main
    run_function_with_points(
  File "/home/runner/work/gis714-2024-tangible-landscape/gis714-2024-tangible-landscape/activities/Gwen.py", line 63, in run_function_with_points
    data = data_lines.split("|")
AttributeError: 'list' object has no attribute 'split'
```
Splitlines() returns list, not string, so you have to iterate through the list and call split on each line.

File: activities/Gwen.py [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+            output="cost_surface",
+        )
+        data = gs.read_command("r.what", map="cost_surface", points="points")
+        data_lines = data.splitlines()
+        connectivity = 0
+        for line in data_lines:
+            line = line.split("|")
+            line = float(line[3])
+            connectivity += 1 / line
+
```
Comment (petrasovaa) [inline activities/Gwen.py]: This entire block needs to be outside of the `if not points` condition

File: activities/Gwen.py [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+        # generate random point for focal garden
+        gs.run_command("v.random", output="focal_garden", npoints=1, seed=3)
+
+        # calculate connectivity between focal garden and points from pins
+        gs.run_command(
+            "r.cost",
+            input="cfactorgrow_1m",
+            start_points="focal_garden",
+            stop_points="points",
```
Comment (petrasovaa) [inline activities/Gwen.py]: You want `points` variable here.
```suggestion
            stop_points=points,
```

**Comment by gkirschke** [comment]
I still don't have a way to display the 'connectivity' value as my output on TL; when I have tried to set up a dashboard before I wasn't getting it right. Does using print() in this interface do anything (even to display the value on the computer for the purposes of this class project)? Otherwise, could someone help me solve the problems I was having with the dashboard?

There seem to be two things happening in the check failures, and I'm not sure how to solve either of them. The render failure can't find the cost surface map that I make in the run_ function, which I thought was available to it. Do I need a placeholder map (in main()?) with that name to display until the run_ function is called? The black issues seem to be with the code in main() that sets the region and makes the test points file. Since I didn't write this code, I'm not sure how I should modify it so that it still runs?

**Comment by petrasovaa** [comment]
> There seem to be two things happening in the check failures, and I'm not sure how to solve either of them. The render failure can't find the cost surface map that I make in the run_ function, which I thought was available to it. Do I need a placeholder map (in main()?) with that name to display until the run_ function is called? The black issues seem to be with the code in main() that sets the region and makes the test points file. Since I didn't write this code, I'm not sure how I should modify it so that it still runs?

See my comments above. Regarding black, just follow what the check suggests (- means what was there before and + means what black wants you to change it to). See also https://github.com/ncsu-geoforall-lab/gis714-2024-tangible-landscape/blob/master/activities/advanced_example.py, the code is aligned to what black suggests.

**Comment by petrasovaa** [comment]
> I still don't have a way to display the 'connectivity' value as my output on TL; when I have tried to set up a dashboard before I wasn't getting it right. Does using print() in this interface do anything (even to display the value on the computer for the purposes of this class project)? Otherwise, could someone help me solve the problems I was having with the dashboard?

You can put the dashboard code you wrote previously in there, but comment it out, because it won't work with the checks, but I can make it run on Tangible Landscape.

File: activities/Gwen.py [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+            height_threshold=[10, 100],
+            cells_threshold=[5, 50],
+            add=True,
+            max_detected=5,
+            debug=True,
+            env=env,
+        )
+
+    # generate random point for focal garden
+    gs.run_command("v.random", output="focal_garden", npoints=1, seed=3)
```
Comment (petrasovaa) [inline activities/Gwen.py]: You need to pass the `env` variable everywhere.

**Comment by gkirschke** [comment]
> You can put the dashboard code you wrote previously in there, but comment it out, because it won't work with the checks, but I can make it run on Tangible Landscape.

Can the dashboard be in the json file? Or will that give an error too, since it won't be able to find it while it is commented out?

**Comment by petrasovaa** [comment]
> > You can put the dashboard code you wrote previously in there, but comment it out, because it won't work with the checks, but I can make it run on Tangible Landscape.
> 
> Can the dashboard be in the json file? Or will that give an error too, since it won't be able to find it while it is commented out?

I believe you can add it to the JSON file. JSON doesn't have comments, so just leave it there.

**Pull request merged**